### PR TITLE
go.mod: switch to Go modules

### PR DESCRIPTION
**Pull request by karalabe**
`+341 −811,974` :stuck_out_tongue_closed_eyes: 

Fixes https://github.com/ethereum/go-ethereum/issues/20226.
Fixes https://github.com/ethereum/go-ethereum/issues/20164.
Fixes https://github.com/ethereum/go-ethereum/issues/20342.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -90,6 +90,10 @@ jobs:
             - python-bzrlib
             - python-paramiko
       script:
+        # Build once locally to download all the go modeule dependencies
+        - go run build/ci.go install
```
Comment (fjl) [inline .travis.yml]: This should be `go mod download`

File: internal/build/util.go [hunk 2]
```diff
@@ -140,6 +140,29 @@ func CopyFile(dst, src string, mode os.FileMode) {
 	}
 }
 
+// CopyFolder copies a folder.
+func CopyFolder(dst, src string, mode os.FileMode) {
```
Comment (fjl) [inline internal/build/util.go]: Probably OK to depend on `github.com/cespare/cp` for this. We already depend on it for something else.

Comment (karalabe) [reply internal/build/util.go]: Switched both CopyFile and CopyFolder over. Does this look ok?

File: .travis.yml (hunk 1 above)
Comment (karalabe) [reply .travis.yml]: I've just tried it, but it downloads a whole lot more (116MB for `go install ./...` vs 156MB for `go mod download`). I guess the tests also depend on a lot of junk. Launchpad will need only the stuff `go install` needs, so might as well leave it as is.

Comment (karalabe) [reply .travis.yml]: Actually `go install -n ./...` does the downloads but omits the actual compilation, so should be the best of both worlds.

Comment (karalabe) [reply .travis.yml]: Added it as a workaround. Not sure if it works, need to merge the pr for it.

**Comment by r8d8** [comment]
https://github.com/ethereum/go-ethereum/pull/19605

**Comment by fjl** [comment]
One thing I don't like about this change is that the debian source package content is no longer deterministic. As written, build/ci.go packages the entire go module cache, which could contain
arbitrary modules checked out locally. I think it would be better to move the invocation of `go install -n ./...` into ci.go and run it with a different cache directory, then copy that directory instead.

Alternatively, we could run `go list` to find the list used modules and only copy those.

**Comment by karalabe** [comment]
I assume Travis' go cache is empty initially so we surely only ship what's
needed. Executing in a diff environment might break this invariant though
indeed.

**Pull request merged**